### ISSUE TITLE
Support Go byte arrays <--> CBOR byte string (major type 2)

### ISSUE THREAD

**Issue by ZenGround0** [fxamacker/cbor#133]
This is a minor request for something that would be mildly helpful if it is not too much trouble.  When attempting to encode a 32 byte array I find that it is unsupported: `unsupported type for encoding: [32]uint8`.

Not much trouble in using a wrapper type with Binary marshaling but flagging that this would be useful.

**Comment by fxamacker** [comment]
I'm having trouble reproducing this issue with v2.0:

```     
	var arr [32]uint8
	for i := 0; i < len(arr); i++ {
		arr[i] = uint8(i)
	}

        // Using Marshal
	b, err := cbor.Marshal(arr)
	fmt.Printf("b: 0x%x, err: %v\n", b, err)

        // Using encoder.Encode
	var buf bytes.Buffer
	enc := cbor.NewEncoder(&buf)
	err = enc.Encode(arr)
	fmt.Printf("b: 0x%x, err: %v\n", buf.Bytes(), err)
```

Output:
```
b: 0x5820000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f, err: <nil>
b: 0x5820000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f, err: <nil>
```
Can you provide a reproducer? 

BTW, thanks for finding the time to file issues and providing feedback!

**Comment by ZenGround0** [comment]
Ah thanks I'm still on v1, feel free to close down.  When I get a chance to upgrade I'll reopen if there's still an issue.

**Comment by fxamacker** [comment]
Just tried it out on release-1.5 branch, and got same results as v2.0.

Here is the test code for v1.5.
```
	var arr [32]uint8
	for i := 0; i < len(arr); i++ {
		arr[i] = uint8(i)
	}
	b, err := cbor.Marshal(arr, cbor.EncOptions{})
	fmt.Printf("b: 0x%x, err: %v\n", b, err)

	var buf bytes.Buffer
	enc := cbor.NewEncoder(&buf, cbor.EncOptions{})
	err = enc.Encode(arr)
	fmt.Printf("b: 0x%x, err: %v\n", buf.Bytes(), err)
```
Output:
```
b: 0x5820000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f, err: <nil>
b: 0x5820000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f, err: <nil>
```
If you were encoding [32]uint8 differently, please let me know.

**Comment by ZenGround0** [comment]
Upon further inspection this is totally the fault of a wrapper we are using.  Sorry for the noise!

**Comment by ZenGround0** [comment]
Reopening with a bit more info.

I also generate a correct encoding in the above case.  The issue I run into comes during decode which raises the error `cbor: cannot unmarshal byte string into Go value of type [32]uint8`

Here is a minimal repeatable example (on v1.5.0, haven't tried on v2.0):

```golang
func TestAliasedByteArrayWrapped(t *testing.T) {
	a := [1]byte{0xfc}
	// serialize and deserialize aliased bytes alone
	aData, err := cbor.Marshal(a, cbor.EncOptions{})
	assert.NoError(t, err)
	fmt.Printf("a bytes: %x\n", aData) // prints 41fc as expected
	var newA [1]byte
	err = cbor.Unmarshal(aData, &newA)
	assert.NoError(t, err) // unexpected error: "cbor: cannot unmarshal byte string into Go value of type [1]uint8"

}
```

**Comment by x448** [comment]
This feature request is for:
* Go byte arrays <--> CBOR byte string (major type 2).

The library currently supports:

* Go arrays <--> CBOR arrays (major type 4).  When array sizes are different, this library handles it like `encoding/json` does in comparable situations.

Go stdlib for encoding/json says,
>To unmarshal a JSON array into a Go array, Unmarshal decodes JSON array elements into corresponding Go array elements. If the Go array is smaller than the JSON array, the additional JSON array elements are discarded. If the JSON array is smaller than the Go array, the additional Go array elements are set to zero values. 

I'm in favor of adding this (unless it causes delays to v2.1 or v2.2).  @fxamacker there might be edge cases or decisions that could benefit from feedback/consensus before implementation.

**Comment by fxamacker** [comment]
This feature is implemented locally but I'm still fuzzing v2.1 -- it's at 1.36 billion exects now.  

It isn't "officially" part of the commit process but I like to fuzz commits (at least briefly) before pushing to GitHub so this will have to wait a bit before landing on GitHub.
